### ISSUE TITLE
breakpoint are skipped whilst using debugger with folder toolchain overrides

### ISSUE THREAD

**Issue by C0D3-M4513R** [intellij-rust/intellij-rust#8636]
## Environment

* **IntelliJ Rust plugin version:** 0.4.159.4232-212
* **Rust toolchain version:** 1.58.1 (db9d1b20b 2022-01-20) x86_64-pc-windows-msvc
* **IDE name and version:** CLion 2021.2.3 (CL-212.5457.51)
* **Operating system:** Windows 10 10.0
* **Macro expansion engine:** new
* **Name resolution engine:** new

## Problem description
Adding a folder toolchain override completly breaks the debugger. Pausing and Breakpoints do not work.

## Steps to reproduce

1. Create a "Hello World" Project (you may want to add a long loop for part 7)
2. Set a Breakpoint on the println! statment.
3. Open a terminal, and select a different toolchain for that folder using the command: `rustup override set [toolchhain]`
4. Add a run configuration, if not present (make sure to set the channel to [default])
5. Hit debug.
6. The programm skips the breakpoint from step 2.
7. Try hitting debug and then hitting pause programm.
8. The debugger will open a disassembly tab. Depending on the project (or when you pause? idk) there may or may not be content in that window.

As the [toolchain] in step 3 I used nightly-i686 and nightly.
My programm I created for testing this is the following:
```rust
fn main() {
	println!("Hello World!");
	let mut i =0;
	while i < 200 {
		println!("{}",i);
		i+=1;
	}
	println!("Hello World!");

	println!("Hello World!");
}
```
Possibly related #7748

**Comment by C0D3-M4513R** [comment]
Breakpoints not getting hit may be releated to #8526. But the broken variable watch remains.

**Comment by Undin** [comment]
I don't think that it's because of toolchain override. Also, I suppose it's not connected to #7748

I think it's just because of using nightly toolchain for compilation and it's a duplicate of #8526. And it's easy to check - just set nightly toolchain as default one and override toolchain for the corresponding directory via toolchain override. I think debugging will work as expected.

Also, see https://github.com/intellij-rust/intellij-rust/issues/8526#issuecomment-1059759448 - debugging with nightly toolchain is fixed in latest CLion EAP

**Comment by C0D3-M4513R** [comment]
The broken debugger and not being able to pause is totally because of #8526.
> just set nightly toolchain

Can't. I do not have the source installed for any other chain exept for stable-x86_64. I also do not know how to do so.

When using stable-i686, which should not have any of the issues regarding #8526 I cannot see any variable contents. So folder toolchain overrides do cause some issues.

**Comment by C0D3-M4513R** [comment]
![grafik](https://user-images.githubusercontent.com/28912031/156943248-1005f707-c3bf-420c-a0b7-2926f089059d.png)

**Comment by Undin** [comment]
Do I understand correctly that:
- you need to compile your rust program to 32 bit executable
- you try to use `+toolchain` mechanism to compile the necessary binary
- the debugger doesn't show values of local variable
?

**Comment by C0D3-M4513R** [comment]
TLDR: Yes. That is correct.
---------------------
Longer version with background:

That is mostly correct. I am writing a library for injecting dlls into other processes (still unreleased, but working mostly).I need to test a specific edge case. In that case my library would need to be compiled as a 32-bit executable. I use the rustup folder override mechanism to achive this, since that allows me to debug my rust code, contrary to temporary toolchain overrides in the run config.

Getting a 32-bit application running under 64-bit is a little bit of magic. Windows impements that magic.
I am basically trying to get a 64bit system call in a 32bit application. To do that I need to undo some of window's magic. Google "Heaven's Gate windows" for more information on magic. But this whole magic cannot be undone, if it doesn't already exist. The magic doesn't exist if your application is 64bit.

**Comment by Undin** [comment]
@C0D3-M4513R thanks for the clarification!

I've managed to reproduce the issue with 32-bit binaries without using rustup override at all. So I want to simplify the issue description (drop redundant steps, for example). If you have any objections, feel free to write about it.

Some related notes:
- > contrary to temporary toolchain overrides in the run config.

  Are you about #7748? If so, then it was fixed in #7864 and the fix will be included in the next release. But you can use [nightly plugin version](https://plugins.jetbrains.com/plugin/8182-rust/docs/rust-quick-start.html#install-nightly) to get it right now. Note, the latest plugin versions support only 2021.3 and 2022.1 major CLion versions. You use 2021.2
  
- you don't need to override toolchain to compile 32bit binaries. Just install the proper build target (`rustup target add i686-pc-windows-msvc`) and pass it in run configuration (add `--target i686-pc-windows-msvc` to command). It will allow you to debug your code without rustup override with your current IDE and plugin versions

**Comment by C0D3-M4513R** [comment]
I updated to the nightly version of the rust plugin. Using temporary toolchain overrides I do not get the same issue.
Cross-Compiling to i686-pc-windows-msvc using stable-x86_x64-pc-windows-msvc causes the values of variables to disappear. Rust folder overrides apparently also cause the values to disappear.
I do not quite know, where the common denominator between the two is.

**Comment by C0D3-M4513R** [comment]
I changed the description.

I just learned, that the `--target` option exists. To me it looks like it cross compiles to x86 with a x64 compiler (when using a x86_64 toolchain), rather than just using a native x86 toolchain. I do not know where the differences lie, if any. I will continue to use temporary toolchain overrides, since they seem more stable for the moment.
